File: src/contract_parse/ast.rs
use std::mem::{MaybeUninit, transmute};

pub struct Contract {
    pub spaces: Vec<SpaceDecl>,
    pub agents: Vec<AgentDecl>,
    pub paths: Vec<PathDecl>,
}

impl Contract {
    pub fn new() -> Contract {
        Contract {
            spaces: vec![],
            agents: vec![],
            paths: vec![],
        }
    }
}

pub struct SpaceDecl {
    // parent: Option<&'a SpaceDecl<'a>>, // todo these lifetimes are weird again
    pub ident: String,
    pub space_type: SpaceType,
    pub params: Vec<Param>,
    pub vibe_desc: VibeBlock,

    // todo space content -- tasks etc.

    pub line_start: usize,
    pub line_end: usize,
}

pub enum SpaceType {
    UI,
    IO,
    DATA,
    FUNC,
    AGENTIC,
    Unknown,
}

pub struct AgentDecl {
    agent_type: AgentType,
    // todo!
}

pub enum AgentType {
    AF,
    DF,
}

pub struct PathDecl {
    path_type: PathType,
    // todo!
}

pub enum PathType {
    READ,
    WRITE,
}

pub struct TaskDecl {
    // todo!
}

pub struct DataDecl {
    //todo!
}

pub struct Param {
    name: *const str,
    binds: *const str,

    mem_source: Option<String>, // in decls, params own their own string memory; in vibe blocks, it is in the vibe prose.

    pub line: usize,
    pub col: usize,
}

impl Param {
    pub fn new(name: *const str, binds: *const str, line: usize, col: usize) -> Param {
        Param {
            name,
            binds,
            mem_source: None,
            line,
            col,
        }
    }

    pub fn new_owned_mem(name: &str, binds: &str, line: usize, col: usize) -> Param {
        let mut p = Param {
            name,
            binds,
            mem_source: Some(format!("{}{}", name, binds)),
            line,
            col,
        };

        // it would be nice to paritially init these because it is an extra memcpy ... oh well ! maybe it will get optimized away
        p.name = &p.mem_source.as_ref().unwrap()[..name.len()];
        p.binds = &p.mem_source.as_ref().unwrap()[name.len()..];

        return p;
    }

    pub fn print(&self) {
        println!("{} = %{}", unsafe { &*self.name }, unsafe { &*self.binds });

    }
}

pub struct VibeBlock {
    vibe_prose: String,
    pub meta_refs: Vec<MetaRef>,

    line_start: usize,
    line_end: usize,
}

pub struct MetaRef {
    ident: *const str,
    ref_info: MetaRefData,

    line: usize,
    col: usize,
}

enum MetaRefData {
    DataRef,
    UseImportSpace(*const str),
    UseImportAgent(*const str),
    TaskRef(Vec<Param>),
    PathRef,
}

impl VibeBlock {
    pub fn new(vibe_prose: String, line_start: usize, line_end: usize) -> VibeBlock {
        let mut block = VibeBlock {
            vibe_prose,
            meta_refs: vec![],
            line_start,
            line_end,
        };

        return block;
    }

    pub fn get_prose(&self) -> &str {
        self.vibe_prose.as_str()
    }
}

impl MetaRef {
    pub fn get_ident(&self) -> &str {
        unsafe {&*self.ident}
    }
}

File: internal/parse/ast.go
package parse

import (
	"strings"
)

type Contract struct {
	spaces []SpaceDecl
	agents []AgentDecl
	paths []PathDecl
}

type SpaceDecl struct {
	ident string
	space_type SpaceType
	replicable bool
	params []Param
	vibe_desc VibeBlock

	// inner decls
	agents []AgentDecl
	tasks []TaskDecl
	// data []DatumDecl

	line_start, line_end uint64
}

type SpaceType byte
const (
	UnknownSpace SpaceType = iota
	UI
	IO
	DATA
	FUNC
	AGENTIC
)

type AgentDecl struct {
	ident string
	agent_type AgentType
	params []Param
	vibe_desc VibeBlock

	line_start, line_end uint64
}

type AgentType byte
const (
	UnknownAgent AgentType = iota
	AF
	DF
)

type PathDecl struct {
	ident string
	path_type PathType
	space_source string
	space_dest string
	vibe_desc VibeBlock

	line_start, line_end uint64
}

type PathType byte
const (
	UnknownPath PathType = iota
	INVOKE
	ATTEND
)

type TaskDecl struct {
	ident string
	params []Param
	vibe_desc VibeBlock

	line_start, line_end uint64
}

// type DatumDecl struct {
// 	ident string
// 	line uint64
// }

type Param struct {
	in_param bool
	data_name string

	line, col uint64
}

func (p *Param) ToStr() string {
	if p.in_param {
		return "in=%" + p.data_name
	} else {
		return "out=%" + p.data_name
	}
}

type VibeBlock struct {
	vibe_prose []string
	meta_refs []MetaRef

	line_start, line_end uint64
}

// can do meta_ref.(type) to get type
type MetaRef interface {
	// Line() int
	// Col() int
	ToStr() string
}

type MetaRefData struct {
	ident string

	line, col uint64
}

func (mr *MetaRefData) ToStr() string {
	return "%" + mr.ident
}

type UseImportType byte
const (
	UseImportSpace UseImportType = iota
	UseImportAgent
)

type MetaRefUseImport struct {
	imported string

	import_type UseImportType

	line, col uint64
}

func (mr *MetaRefUseImport) ToStr() string {
	var joiner string
	switch mr.import_type {
	case UseImportSpace: joiner = "@"
	case UseImportAgent: joiner = "#"
	default: panic(-1)
	}
	return "$use(" + joiner + mr.imported + ")"
}

type MetaRefTask struct {
	ident string

	line, col uint64

	args []Param
}

func (mr *MetaRefTask) ToStr() string {
	arg_strs := make([]string, len(mr.args), len(mr.args))
	for i := 0; i < len(mr.args); i++ {
		arg_strs[i] = mr.args[i].ToStr()
	}
	return "$" + mr.ident + "(" + strings.Join(arg_strs, ", ") + ")"
}

type MetaRefPath struct {
	ident string

	line, col uint64
}

func (mr *MetaRefPath) ToStr() string {
	return "=" + mr.ident
}
